### ISSUE TITLE
Bug in fragment caching: object reachable from multiple relationships

### ISSUE THREAD

**Issue by drakedowsett** [JSONAPI-Resources/jsonapi-resources#803]
Seeing an issue introduced by #712 on `master` branch; similar testing app to #802, with additional thing-to-thing reference:

```
                  /----\ many
boxes ------> things <-/
       many    | ^
           one | |
               | | many
               v |
              users
```

With an addition of a relationship from `Thing` 1 to `Thing` 2, I can request for things-to-user:

```
# /api/boxes?include=things,things.user
...
    {
      "id": "1",
      "type": "things",
      "relationships": {
        "user": {
          "data": { "type": "users", "id": "1" }
        },
        "things": {}
      }
    },
    {
      "id": "2",
      "type": "things",
      "relationships": {
        "user": {
          "data": { "type": "users", "id": "1" }
        },
        "things": {}
      }
    },
    {
      "id": "1",
      "type": "users",
      "relationships": {
        "things": {}
      }
    }
...
```

And I can request for things-to-things:

```
# /api/boxes?include=things,things.things
...
    {
      "id": "1",
      "type": "things",
      "relationships": {
        "user": {},
        "things": {
          "data": [
            { "type": "things", "id": "2" }
          ]
        }
      }
    },
    {
      "id": "2",
      "type": "things",
      "relationships": {
        "user": {},
        "things": {}
      }
    },
...
```

But when I request both, `Thing` 2 is missing its thing-to-user data:

```
# /api/boxes?include=things,things.user,things.things
...
    {
      "id": "1",
      "type": "things",
      "relationships": {
        "user": {
          "data": { "type": "users", "id": "1" }
        },
        "things": {
          "data": [
            { "type": "things", "id": "2" }
          ]
        }
      }
    },
    {
      "id": "2",
      "type": "things",
      "relationships": {
        "user": {},
        "things": {}
      }
    },
    {
      "id": "1",
      "type": "users",
      "relationships": {
        "things": {}
      }
    }
...
```

Note that I am not using the `caching` directive in my resources, so I would not expect any change in #712 to take effect. Seems like an unintended side effect.

Additional rails app code:

```
class Thing < ActiveRecord::Base
  belongs_to :box
  belongs_to :user

  # added:
  has_many :related_things, foreign_key: :from_id
  has_many :things, through: :related_things, source: :to
end

class RelatedThing < ActiveRecord::Base
  belongs_to :from, class_name: Thing, foreign_key: :from_id
  belongs_to :to, class_name: Thing, foreign_key: :to_id
end

class Api::ThingResource < JSONAPI::Resource
  has_one :box
  has_one :user

  # added:
  has_many :things
end
```

**Comment by DavidMikeSimon** [comment]
@drakedowsett Can you provide the complete example app and data you used to demonstrate this bug?

**Comment by drakedowsett** [comment]
@DavidMikeSimon here you go: https://github.com/drakedowsett/jsonapi-resources-issues/tree/issue-1

Have tried to keep it as vanilla as possible. Please let me know if you need anything else.

**Comment by DavidMikeSimon** [comment]
@drakedowsett Thank you very much. :-) Could you also add a test to that repository which fails due to the bug?

**Comment by drakedowsett** [comment]
@DavidMikeSimon update with a spec that fails on the ref-merge of #712 but passes on tag `v0.8.0.beta2`.

**Comment by DavidMikeSimon** [comment]
Sorry about the delay on this, i promise i haven't forgotten this bug.

**Comment by lgebhardt** [comment]
@DavidMikeSimon Yesterday I merged #832, which I believe fixes this as well. It looks like there were really two issues against this same problem and I only marked #802 as closed.

Can we confirm that this has been fixed on the latest master branch and in the 0.8.0 release?

**Comment by drakedowsett** [comment]
@lgebhardt @DavidMikeSimon This is also addressed! (see note on #802). I would add though that even if `deep_merge` was the solution for both, this second issue seemed to be specifically related to #712 as it wasn't an issue before that PR was merged into master. Thanks both of you!

**Comment by lgebhardt** [comment]
@drakedowsett Thanks for confirming the fix. The problem was one of order. Resources are serialized by walking down the primary resources and serializing their relationships that are specified in the includes. When a resource is encountered more than one time it was not serialized again. This was fine as long as the specified includes were the same (so it was serialized the same) however in some cases the includes were not the same. The results were variable and depended on the order the resources were serialized. My guess is #712 didn't actually cause the issue, it just rendered resources in a different order so it appeared it did.
